doc: support the view hint (#11466)
pingcap/docs: optimizer-hints.md

>
> When the number of CTE references is too high, the query performance might be lower than the default materialization behavior.

## Hints that take effect globally

> The global hint is related to [views](/views.md), and enables hints defined in the query to be effective inside the view. To specify global hints, you can first use the `QB_NAME` hint to define a query block name, and then add hints in the form of `ViewName@QueryBlockName`.

Author: ```suggestion
The global hint works in [views](/views.md). When specified as a global hint, the hint defined in a query can take effect inside the view. To specify a global hint, first use the `QB_NAME` hint to define a query block name, and then add the target hints in the form of `ViewName@QueryBlockName`.
```

### Step 1: Define the query block name of the view using the `QB_NAME` hint

You can use the [`QB_NAME` hint](#qb_name) to define a new name for each query block of the view. The concept of the `QB_NAME` hint in the view is the same as that in [query blocks]((#qb_name)), but the syntax is extended from `QB_NAME(QB)` to `QB_NAME(QB, ViewName@QueryBlockName [.ViewName@QueryBlockName .ViewName@QueryBlockName ...])`.

> **Note:**
>
> There is a white space between `@QueryBlockName` and the immediately followed `.ViewName@QueryBlockName`. Otherwise, the `.ViewName@QueryBlockName` will be treated as a part of the `QueryBlockName`. For example, `QB_NAME(v2_1, v2@SEL_1 .@SEL_1)` is valid, while `QB_NAME(v2_1, v2@SEL_1.@SEL_1)` cannot be parsed correctly.

> - For a simple statement with a single view and no subqueries, the following example specifies the first query block name of view `v`:

Reviewer: ```suggestion
- For a simple statement with a single view and no subqueries, the following example specifies the first query block rename of view `v`:
```

    ```sql

>     SELECT /* The name of the current query block is the default @SEL_1 */ * FROM v;

Reviewer: ```suggestion
    SELECT /* Comment: The name of the current query block is the default @SEL_1 */ * FROM v;
```

    ```

    For view `v`, the first view name in the list (`ViewName@QueryBlockName [.ViewName@QueryBlockName .ViewName@QueryBlockName ...]`) starting from the query statement is `v@SEL_1`. The first query block of the view `v` can be declared as `QB_NAME(v_1, v@SEL_1 .@SEL_1)`, or simply written as `QB_NAME(v_1, v)`, omitting `@SEL_1`:

    ```sql

>     CREATE VIEW v AS SELECT /* The name of the current query block is the default @SEL_1 */ * FROM t;

Reviewer: ```suggestion
    CREATE VIEW v AS SELECT /* Comment: The name of the current query block is the default @SEL_1 */ * FROM t;
```

>     -- Specify the global Hint

Reviewer 2: ```suggestion
    -- Specifies the global hint
```

    SELECT /*+ QB_NAME(v_1, v) USE_INDEX(t@v_1, idx) */ * FROM v;
    ```

> - For a complex statement with nested views and subqueries, the following example specifies the names for each two query blocks of the view `v1` and `v2`:

Reviewer 2: ```suggestion
- For a complex statement with nested views and subqueries, the following example specifies the names for each of two query blocks of the view `v1` and `v2`:
```

    ```sql
    SELECT /* The name of the current query block is the default @SEL_1 */ * FROM v2 JOIN (

>         SELECT /* The name of the current query block is the default @SEL_2 */ * FROM v2) vv;

Reviewer: ```suggestion
        SELECT /* Comment: The name of the current query block is the default @SEL_2 */ * FROM v2) vv;
```

    ```

    For the first view `v2`, the first view name in the list starting from the first query statement is `v2@SEL_1`. For the second view `v2`, the first view name is `v2@SEL_2`. The following example only considers the first view `v2`.

    The first query block of view `v2` can be declared as `QB_NAME(v2_1, v2@SEL_1 .@SEL_1)`, and the second query block of the view `v2` can be declared as `QB_NAME(v2_2, v2@SEL_1 .@SEL_2)`:

    ```sql
    CREATE VIEW v2 AS

>         SELECT * FROM t JOIN /* For view v2, the name of the current query block is the default @SEL_1. So, the current query block view list is v2@SEL_1 .@SEL_1 */

Reviewer: ```suggestion
        SELECT * FROM t JOIN /* Comment: For view v2, the name of the current query block is the default @SEL_1. So, the current query block view list is v2@SEL_1 .@SEL_1 */
```

        (

>             SELECT COUNT(*) FROM t1 JOIN v1 /* For view v2, the name of the current query block is the default @SEL_2. So, the current query block view list is v2@SEL_1 .@SEL_2 */

Reviewer: ```suggestion
            SELECT COUNT(*) FROM t1 JOIN v1 /* Comment: For view v2, the name of the current query block is the default @SEL_2. So, the current query block view list is v2@SEL_1 .@SEL_2 */
```

        ) tt;
    ```

    For view `v1`, the first view name in the list starting from the preceding statement is `v2@SEL_1 .v1@SEL_2`. The first query block in view `v1` can be declared as `QB_NAME(v1_1, v2@SEL_1 .v1@SEL_2 .@SEL_1)`, and the second query block in view `v1` can be declared as `QB_NAME(v1_2, v2@SEL_1 .v1@SEL_2 .@SEL_2)`:

    ```sql

>     CREATE VIEW v1 AS SELECT * FROM t JOIN /* For view `v1`, the name of the current query block is the default @SEL_1. So, the current query block view list is v2@SEL_1 .@SEL_2 .v1@SEL_1 */

Reviewer: ```suggestion
    CREATE VIEW v1 AS SELECT * FROM t JOIN /* Comment: For view `v1`, the name of the current query block is the default @SEL_1. So, the current query block view list is v2@SEL_1 .@SEL_2 .v1@SEL_1 */
```

        (

>             SELECT COUNT(*) FROM t1 JOIN t2 /* For view `v1`, the name of the current query block is the default @SEL_2. So, the current query block view list is v2@SEL_1 .@SEL_2 .v1@SEL_2 */

Reviewer: ```suggestion
            SELECT COUNT(*) FROM t1 JOIN t2 /* Comment: For view `v1`, the name of the current query block is the default @SEL_2. So, the current query block view list is v2@SEL_1 .@SEL_2 .v1@SEL_2 */
```

>
> - When you define the `QB_NAME` hint in a view for the outermost query block:
>
>     - For the first item of the view list in the `QB_NAME`, if the `@SEL_` is not explicitly declared, the default is consistent with the query block position where the `QB_NAME` is defined. That is, the query `SELECT /*+ QB_NAME(qb1, v2) */ * FROM v2 JOIN (SELECT /*+ QB_NAME(qb2, v2) */ * FROM v2) vv;` is equivalent to `SELECT /*+ QB_NAME(qb1, v2@SEL_1) */ * FROM v2 JOIN (SELECT /*+ QB_NAME(qb2, v2@SEL_2) */ * FROM v2) vv;`.
>     - For items other than the first item of the view list in the `QB_NAME`, only `@SEL_1` can be omitted. That is, if `@SEL_1` is declared in the first query block of the current view, `@SEL_1` can be omitted. Otherwise, `@SEL_` cannot be omitted. For the preceding example:
>
>         - The first query block of the view `v2` can be declared as `QB_NAME(v2_1, v2)`.
>         - The second query block of the view `v2` can be declared as `QB_NAME(v2_2, v2.@SEL_2)`.
>         - The first query block of the view `v1` can be declared as `QB_NAME(v1_1, v2.v1@SEL_2)`.
>         - The second query block of the view `v1` can be declared as `QB_NAME(v1_2, v2.v1@SEL_2 .@SEL_2)`.

> ### Step 2: Add required hints

Author: ```suggestion
### Step 2: Add the target hints
```

> After defining the `QB_NAME` hint for query blocks of the view, you can add required [hints that take effect in query blocks](#hints-that-take-effect-in-query-blocks) in the form of `ViewName@QueryBlockName` to make them take effect inside the view. For example:

Reviewer 2: ```suggestion
After defining the `QB_NAME` hint for query blocks of the view, you can add required [hints that take effect in query blocks](#hints-that-take-effect-in-query-blocks) in the form of `ViewName@QueryBlockName` to make them effective inside the view. For example:
```
